### ISSUE TITLE
Getting the i'th smallest or largest element of a BTreeSet

### ISSUE THREAD

**Issue by dfoxfranke** [rust-lang/rust#53318]
I'm in need of a data structure implementing an ordered set in which the following operations are all O(log N):

* Insertion
* Deletion
* Looking up the i'th smallest or largest element

B-trees should lend themselves well to this sort of lookup, but `BTreeSet` does not currently seem to support it.

**Comment by sfackler** [comment]
I don't think that can be done without adding extra metadata to the node type which we probably wouldn't want to do to support an obscure operation.

**Comment by sfackler** [comment]
They'd run in `ilog(n)` time, not `log(n)`

**Comment by ExpHP** [comment]
`skip` or `nth` could be overriden, though, and judging from the source it appears that subtree lengths are available... so I would imagine/hope that this can be made `log(n)`?

**Comment by abonander** [comment]
That's in the worst case though, isn't it? In practice you'd only have the logarithmic cost on the first iteration, right? After that it's visiting sibling items and nodes.

**Comment by dfoxfranke** [comment]
I'd expect it to be O(i + log N), which is O(N) for the worst case where i=N/2.

**Comment by sfackler** [comment]
> judging from the source it appears that subtree lengths are available

I don't think that's stored - `InternalNode` has a len field, but that's the number of values in that node itself, not the number of values in the subtree as a whole.

> That's in the worst case though, isn't it? In practice you'd only have the logarithmic cost on the first iteration, right? After that it's visiting sibling items and nodes.

Iteration requires traversing the tree AFAIKT.

**Comment by dfoxfranke** [comment]
@sfackler fair enough if it's going to require additional metadata. I'll fork the implementation into a separate package if this gets to be a pressing issue.

**Comment by abonander** [comment]
> Iteration requires traversing the tree AFAIKT.

Yes but most of that cost looks to be paid up-front when you call `iter()`; it immediately descends to and memoizes the lowest and highest valued leaf nodes. From there iteration yields the sibling items in the node (effectively constant time) and then ascends into the parent node and down into the next child in order which is `O(log(n))` in the worst case if you have to go all the way up the tree to the root and down the right side but at minimum only 2 pointer traversals (self to parent to child).

**Comment by Dylan-DPC** [comment]
Closing this as it's better suited for a crate to provide this data structure instead of expecting rustc to have it

**Comment by kaathewise** [comment]
> Iteration requires traversing the tree AFAIKT.

FYI, this is incorrect: the existing metadata is enough to implement O(log(N)) indexing and finding index by element, because `Node.len` field allows you to skip the entire subtree without traversing it.
